web_src/js/modules/tippy.js

```diff
@@ -25,38 +25,108 @@ export function createTippy(target, opts = {}) {
+  if (!content && target.classList.contains('tooltip')) {
+    content = target.getAttribute('data-content');
+  }
+  return content;
+}
+
+/**
+ * Attach a tippy tooltip to the given target element.
+ * If the target element already has a tippy tooltip attached, the tooltip will be updated with the new content.
+ * If the target element has no content, then no tooltip will be attached, and it returns null.
+ * @param target {HTMLElement}
+ * @param content {null|string}
+ * @returns {null|tippy}
+ */
+function attachTippyTooltip(target, content = null) {
+  content = content ?? getTippyTooltipContent(target);
   if (!content) return null;
-  if (!el.hasAttribute('aria-label')) el.setAttribute('aria-label', content);
-  return createTippy(el, {
+
+  const props = {
     content,
     delay: 100,
     role: 'tooltip',
-    ...(el.getAttribute('data-tooltip-interactive') === 'true' ? {interactive: true} : {}),
-    ...props,
-  });
-}
+    placement: target.getAttribute('data-tooltip-placement') || 'top-start',
+    ...(target.getAttribute('data-tooltip-interactive') === 'true' ? {interactive: true} : {}),
+  };
 
-export function showTemporaryTooltip(target, content) {
-  let tippy, oldContent;
-  if (target._tippy) {
-    tippy = target._tippy;
-    oldContent = tippy.props.content;
+  if (!target._tippy) {
+    createTippy(target, props);
   } else {
-    tippy = initTooltip(target, {content});
+    target._tippy.setProps(props);
+  }
+  return target._tippy;
+}
+
+/**
+ * creating tippy instance is expensive, so we only create it when the user hovers over the element
+ * @param e {Event}
+ */
+function lazyTippyOnMouseEnter(e) {
+  e.target.removeEventListener('mouseenter', lazyTippyOnMouseEnter, true);
+  attachTippyTooltip(this);
+}
+
+/**
+ * Activate the tippy tooltip for all children elements
+ * And if the element has no aria-label, use the tooltip content as aria-label
+ * @param target {HTMLElement}
+ */
+function attachChildrenLazyTippyTooltip(target) {
+  // the selector must match the logic in getTippyTooltipContent
+  for (const el of target.querySelectorAll('[data-tooltip-content], .tooltip[data-content]')) {
+    el.addEventListener('mouseenter', lazyTippyOnMouseEnter, true);
+
+    // meanwhile, if the element has no aria-label, use the tooltip content as aria-label
+    if (!el.hasAttribute('aria-label')) {
+      const content = getTippyTooltipContent(el);
+      if (content) {
+        el.setAttribute('aria-label', content);
+      }
+    }
   }
+}
 
+export function initGlobalTooltips() {
+  // use MutationObserver to detect new elements added to the DOM, or attributes changed
+  const observer = new MutationObserver((mutationList) => {
+    for (const mutation of mutationList) {
+      if (mutation.type === 'childList') {
+        for (const el of mutation.addedNodes) {
+          // handle all "tooltip" elements in newly added nodes, skip non-related nodes (eg: "#text")
+          if (el.querySelectorAll) {
```

Comment:
this line looks broken.

Reply:
Not broken, just a quick check to see the "node" has `querySelectorAll` function.

Otherwise some nodes (like `#text`) couldn't be used to attach tippy.

Reply:
Use [node.nodeType](https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType) please, it's much more clear.

Reply:
Should be `[Node.ELEMENT_NODE,Node.DOCUMENT_FRAGMENT_NODE].includes(el.nodeType)`.

Reply:
OK, then `'querySelectorAll' in el` : ba8ad89e5148e89dddb0789a4dac6f380197cb83

Reply:
Ok